From d5b68904caf821241a22bc057ab30840654e4531 Mon Sep 17 00:00:00 2001
From: Jack Ping CHNG <jack.ping.chng@linux.intel.com>
Date: Mon, 20 Apr 2020 17:34:30 +0800
Subject: [PATCH] DRV_DPM_SW-229: DPM remove dp_dev_list_free

- removed dp_dev_list_free as it is not in used
---
 drivers/net/datapath/dpm/datapath_instance.c | 28 +++-------------------------
 1 file changed, 3 insertions(+), 25 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_instance.c b/drivers/net/datapath/dpm/datapath_instance.c
index 59b2a9cd4d46..3c5a949fe917 100644
--- a/drivers/net/datapath/dpm/datapath_instance.c
+++ b/drivers/net/datapath/dpm/datapath_instance.c
@@ -14,14 +14,8 @@
 int dp_cap_num;
 struct dp_hw_cap hw_cap_list[DP_MAX_HW_CAP];
 
-/*Subif DEV hash list
- *Note: in current implementation, its memory will not be really removed,
- *Instead, it is just move to a free hash list dp_dev_list_free
- *Reason: if we really free the memory, net_device's opt callback sometimes
- *will be set to NULL?? Need further check
- */
+/*Subif DEV hash list */
 struct hlist_head dp_dev_list[DP_DEV_HASH_SIZE];
-struct hlist_head dp_dev_list_free[DP_DEV_HASH_SIZE];
 
 /*Module hash list */
 struct hlist_head dp_mod_list[DP_MOD_HASH_SIZE];
@@ -259,16 +253,6 @@ int dp_inst_add_dev(struct net_device *dev, char *subif_name, int inst,
 
 	subif->subif = ctp;
 	dp_dev = dp_dev_lookup(&dp_dev_list[idx], dev, subif_name, flag);
-
-	if (!dp_dev) { /*search free list */
-		dp_dev = dp_dev_lookup(&dp_dev_list_free[idx],
-				       dev, subif_name, flag);
-		if (dp_dev) {
-			hlist_del(&dp_dev->hlist); /*remove from free list */
-			dp_dev->count = 0;
-			new_f = 1;
-		}
-	}
 	if (!dp_dev) { /*alloc new */
 		dp_dev = devm_kzalloc(&g_dp_dev->dev, sizeof(*dp_dev),
 				      GFP_ATOMIC);
@@ -369,11 +353,7 @@ int dp_inst_del_dev(struct net_device *dev, char *subif_name, int inst,
 				hlist_del(&dp_dev->hlist);
 				if (dev)
 					dp_ops_reset(dp_dev, dev);
-				/*do't really free now
-				 *in case network stack is holding the callback
-				 */
-				hlist_add_head(&dp_dev->hlist,
-					       &dp_dev_list_free[idx]);
+				devm_kfree(&g_dp_dev->dev, dp_dev);
 			}
 			return 0;
 		}
@@ -673,10 +653,8 @@ int dp_inst_init(u32 flag)
 
 	dp_cap_num = 0;
 	memset(hw_cap_list, 0, sizeof(hw_cap_list));
-	for (i = 0; i < DP_DEV_HASH_SIZE; i++) {
+	for (i = 0; i < DP_DEV_HASH_SIZE; i++)
 		INIT_HLIST_HEAD(&dp_dev_list[i]);
-		INIT_HLIST_HEAD(&dp_dev_list_free[i]);
-	}
 
 	return 0;
 }
